include/roboteam_utils/containers/circular_buffer.h

```diff
@@ -0,0 +1,115 @@
+//
+// Created by rolf on 26-09-20.
+//
+
+#ifndef RTT_CIRCULAR_BUFFER_H
+#define RTT_CIRCULAR_BUFFER_H
+
+#include <stdexcept>
+
+namespace rtt::collections {
+/**
+ * @brief A standard container for Circular buffers.
+ * Note this class does not have iterators because they do not comply with the standard as there would be cases where
+ * first == last. This quickly becomes very ugly.
+ * The front of the buffer are the oldest elements, e.g. the ones first pushed in. These are the next to be erased.
+ * @tparam Tp value type
+ * @tparam Size largest size the buffer can hold
+ */
+    template<typename Tp, std::size_t Size>
+    class circular_buffer {
+    public:
+        typedef Tp value_type;
+        typedef value_type *pointer;
+        typedef const value_type *const_pointer;
+        typedef value_type &reference;
+        typedef const value_type &const_reference;
+        typedef std::size_t size_type;
+
+    private:
+        // the internal array that actually stores the values
+        Tp _m_buff[Size];
+
+        //The virtual beginning of the circular buffer
+        size_type _current_offset;
+
+        //The current size of the circular buffer
+        size_type _current_size;
+
+    public:
+        //Default constructor
+        circular_buffer() : _current_offset(0), _current_size(0) {
+
+        }
+
+        //modifying functions
+        Tp &pop_front() {
+            Tp &ret = _m_buff[_current_offset++];
+            _current_offset %= Size;
```

Comment:
Shoudn't you modulo _after_ the decrementing?

Reply:
I don't understand your question. I increment the offset and then modulo, so indeed yes, I agree with you. it might be confusing now because _Size is compile time constant whereas _current_size is not. This function right now is UB if the length is already zero.